### ISSUE TITLE
[C#] Implement new Concept Exercise: tuples-advanced

### ISSUE THREAD

**Issue by ErikSchierboom** [exercism/v3#964]
This issue describes how to implement the `tuples-advanced` concept exercise for the C# track.

## Goal

The goal of this exercise is to teach the student some of the advanced features of the Concept of Tuples in C#.

## Learning objectives

- Know how to use discards for tuple deconstruction.
- Know how to pattern matching on tuples.
- Know how to add tuple pattern matching to your own and built-in types.
- Know that the `ValueTuple` type has obsoleted the `Tuple` type and why.

## Out of scope

Nothing.

## Concepts

This Concepts Exercise's Concepts are:

- `tuples-advanced`: know how to use discards for tuple deconstruction; know how to pattern matching on tuples; know how to add tuple pattern matching to other types; know that the `ValueTuple` type has obsoleted the `Tuple` type and why.

## Prequisites

This Concept Exercise's prerequisites Concepts are:

- `tuples-basic`: know the basics about tuples.
- `pattern-matching-basic`: know how to do pattern matching.
- `classes-basic`: a user-defined class will have tuple deconstruction added to it.
- `extension-methods`: a built-in library class will have tuple deconstruction added to it.

### Hints

- [Tuple pattern][tuple-pattern]: shows how to pattern match tuples.
- [Deconstructing tuples][deconstructing-tuples]: shows how to do tuple deconstruction for custom and built-in types.

### After

- [Tuple pattern][tuple-pattern]: shows how to pattern match tuples.
- [Deconstructing tuples][deconstructing-tuples]: shows how to do tuple deconstruction for custom and built-in types.

## Representer

This exercise does not require any specific representation logic to be added to the [representer][representer].

## Analyzer

This exercise does not require any specific analyzer logic to be added to the [analyzer][analyzer].

## Implementing

If you'd like to work on implementing this exercise, the first step is to let us know through a comment on this issue, to prevent multiple people from working on the same exercise. If you have any questions while implementing the exercise, please also post them as comments in this issue.

Implementing the exercise means creating the following files:

<pre>
languages
└── csharp
    └── exercises
        └── concept
            └── tuples-advanced
                ├── .docs
                |   ├── after.md
                |   ├── hints.md
                |   ├── instructions.md
                |   └── introduction.md
                ├── .meta
                |   ├── config.json
                |   ├── design.md
                |   └── Example.cs
                ├── TuplesAdvanced.csproj
                ├── TuplesAdvanced.cs
                └── TuplesAdvancedTests.cs
</pre>

## Step 1: add .docs/introduction.md

This file contains an introduction to the concept. It should be explicit about what the student should learn from the exercise, and provide a short, concise introduction to the concept(s). The aim is to give the student just enough context to figure things out themselves and solve the exercise, as research has shown that self-discovery is the most effective learning experience. Mentioning technical terms that the student can Google if they so want, is preferable over including any code samples or an extensive description. For example we might describe a string as a "Sequence of Unicode characters" or a "series of bytes" or "an object". Unless the student needs to understand the details of what those mean to be able to solve the exercise we should not give more info in this introduction - instead allowing the student to Google, ignore, or map their existing knowledge.

## Step 2: add .docs/instructions.md

This file contains instructions for the exercise. It should explicitly explain what the student needs to do (define a method with the signature `X(...)` that takes an A and returns a Z), and provide at least one example usage of that function. If there are multiple tasks within the exercise, it should provide an example of each.

## Step 3: add .docs/hints.md

If the student gets stuck, we will allow them to click a button requesting a hint, which shows this file. This will not be a "recommended" path and we will (softly) discourage them using it unless they can't progress without it. As such, it's worth considering that the student reading it will be a little confused/overwhelmed and maybe frustrated.

The file should contain both general and task-specific "hints". These hints should be enough to unblock almost any student. They might link to the docs of the functions that need to be used.

The hints should not spell out the solution, but instead point to a resource describing the solution (e.g. linking to documentation for the function to use).

## Step 4: add .docs/after.md

Once the student completes the exercise they will be shown this file, which should provide them with a summary of what the exercise aimed to teach. This document can also link to any additional resources that might be interesting to the student in the context of the exercise.

The above four files are also all described in the [concept exercises document][concept-exercises].

## Step 5: update languages/csharp/config.json

An entry should be added to the track's `config.json` file for the new concept exercise:

```json
{
  ...
  "exercises": {
    "concept": [
      ...
      {
        "slug": "tuples-advanced",
        "uuid": "8323dd7c-6aad-11ea-bc55-0242ac130003",
        "concepts": ["tuples-advanced"],
        "prerequisites": [
          "tuples-basic",
          "pattern-matching-basic",
          "classes-basic",
          "extension-methods"
        ]
      }
    ]
  }
}
```

## Step 6: adding track-specific files

These files are specific to the C# track:

- `TuplesAdvanced.csproj`: the C# project file.
- `TuplesAdvancedTests.cs`: the test suite.
- `TuplesAdvanced.cs`. the stub implementation file, which is the starting point for students to work on the exercise.
- `.meta/Example.cs`: an example implementation that passes all the tests.

Check out [an existing exercise][exercise-example] to see what these files should look like.

## Step 7: update the general concept document

Add the exercise to the [concept's shared document's][reference-doc] `## Implementations` section ([example][reference-example]). As there currently is no string formatting shared document, a stub version of that file should be created at `reference/concepts/string_formatting.md`.

## Step 8: updating list of implemented exercises

- Add the exercise to the [list of implemented exercises][implemented-exercises].

## Step 9: add .meta/design.md:

This file contains information on the exercise's design, which includes things like its goal, its teaching goals, what not to teach, and more ([example][design-example]). This information can be extracted from this GitHub issue.

## Step 10: add .meta/config.json:

This file contains meta information on the exercise, which currently only includes the exercise's contributors ([example][config.json-example]).

### Inspiration

When implementing this exericse, it can be very useful to look at [already implemented C# exercises][implemented-exercises].

[how-to-implement-a-concept-exercise]: https://github.com/exercism/v3/blob/master/docs/maintainers/generic-how-to-implement-a-concept-exercise.md
[implemented-exercises]: https://github.com/exercism/v3/tree/master/languages/csharp/exercises/concept/README.md#implemented-exercises
[reference]: https://github.com/exercism/v3/blob/master/languages/csharp/reference/README.md#reference-docs
[reference-doc]: https://github.com/exercism/v3/blob/master/reference/types/tuple.md
[reference-example]: https://github.com/exercism/v3/blob/master/reference/types/string.md#implementations
[analyzer]: https://github.com/exercism/csharp-analyzer
[representer]: https://github.com/exercism/csharp-representer
[exercise-example]: https://github.com/exercism/v3/tree/master/languages/csharp/exercises/concept/numbers-floating-point
[design-example]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/numbers/.meta/design.md
[config.json-example]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/numbers/.meta/config.json
[concept-exercises]: https://github.com/exercism/v3/blob/master/docs/concept-exercises.md
[tuple-pattern]: https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-8#tuple-patterns
[deconstructing-tuples]: https://docs.microsoft.com/en-us/dotnet/csharp/deconstruct

**Comment by mikedamay** [comment]
Consider forking some of [tuples](https://github.com/exercism/v3/tree/master/languages/elixir/exercises/concept/tuples) from the elixir track.
